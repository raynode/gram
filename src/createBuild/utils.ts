import { isType } from 'graphql'
import { isEmpty, map, reduce } from 'lodash'
import { isBuildModeGenerator } from './guards'
import {
  BuildModeGenerator,
  Fields,
  FieldType,
  GQLRecord,
  SimpleFieldType,
} from './types'

export const reduceRecord = <Source, Context>(
  fields: Fields<Source, Context>,
) =>
  reduce(
    fields,
    (list, { args = {}, type }, name) => {
      const parameters = isEmpty(args)
        ? ''
        : `(${map(args, (type, arg) => `${arg}: ${type}`)})`
      list.push(`${name}${parameters}: ${type}`)
      return list
    },
    [],
  ).join('\n    ')

export const createBuildModeResolver = <BuildMode>(buildMode: BuildMode) => <
  Result
>(
  data: Result | BuildModeGenerator<BuildMode, Result>,
) => (isBuildModeGenerator<BuildMode, Result>(data) ? data(buildMode) : data)

export const simpleFieldTypeToFieldType = <Source, Context>(
  simpleType: SimpleFieldType<Source, Context>,
): FieldType<Source, Context> => {
  if (isType(simpleType)) return { type: simpleType.toString() }
  if (typeof simpleType === 'string') return { type: simpleType }
  return simpleType
}

export const convertSimpleFieldsToFields = <Source, Context>(
  simpleFields: Record<string, SimpleFieldType<Source, Context>>,
): Fields<Source, Context> =>
  reduce(
    simpleFields,
    (record, simpleField, name) => {
      record[name] = simpleFieldTypeToFieldType(simpleField)
      return record
    },
    {},
  )
