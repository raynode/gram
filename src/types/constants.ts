export const MODELBUILDER = 'BUILDER::MODEL'
export const ATTRIBUTEBUILDER = 'BUILDER::ATTRIBUTE'
export const SCHEMABUILDER = 'BUILDER::SCHEMA'
export const GQLBUILDER = 'BUILDER:GQLBUILDER'
