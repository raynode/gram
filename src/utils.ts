import {
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLInputFieldConfigMap,
  GraphQLInputObjectType,
  GraphQLInputType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLType,
  isType,
} from 'graphql'
import { identity, map, memoize, pickBy, reduce } from 'lodash'

import { buildType } from './attributeBuilder'
import {
  AttributeBuilder,
  ContextFn,
  ContextModel,
  ContextModelFn,
  DataType,
  ModelBuilder,
  ModelType,
  PageData,
  WithContext,
  Wrapped,
} from './types'

import { Names } from './strategies/naming'

export const isContextFn = <BuildMode, Type>(
  val: any,
): val is ContextFn<BuildMode, Type> => typeof val === 'function'

export const extractData = <BuildMode, Type>(
  data: WithContext<BuildMode, Type>,
) => (buildMode: Wrapped<BuildMode>) =>
  isContextFn(data) ? data(buildMode) : data

export const record = (service: Record<string, any>) => ({
  exists: (key: string) =>
    service.hasOwnProperty(key) && typeof service[key] === 'function',
})

export const clearRecord = (record: Record<string, any>) =>
  pickBy(record, identity)

export type FieldReducerFn<BuildMode> = (
  fields: GraphQLFieldConfigMap<any, any>,
  model: ModelBuilder<BuildMode, any>,
) => GraphQLFieldConfigMap<any, any>

export const fieldsReducer = <BuildMode>(
  reducer: (
    buildModeModel: ContextModel<BuildMode, any>,
  ) => GraphQLFieldConfigMap<any, any> | GraphQLInputFieldConfigMap,
) => (buildMode: Wrapped<BuildMode>): FieldReducerFn<BuildMode> => (
  fields,
  model,
) => ({ ...fields, ...clearRecord(reducer(model.build(buildMode))) })

// this Type construct will ensure that the returned object will have the same keys as
// the input object. It will also convert the properties from ModelBuilder to FieldReducer
export const reduceFields = <
  BuildMode,
  Reducers,
  Models,
  ReducerKeys extends keyof Reducers
>(
  models: Record<string, ModelBuilder<BuildMode, any>>,
  reducers: Record<ReducerKeys, FieldReducerFn<BuildMode>>,
  fields: Record<ReducerKeys, GraphQLFieldConfigMap<any, any>>,
): Record<ReducerKeys, GraphQLFieldConfigMap<any, any>> =>
  reduce(
    reducers,
    (fields, reducer, name) => ({
      ...fields,
      [name]: reduce(models, reducer, fields[name]),
    }),
    fields,
  )

export type ToContextFnResult<BuildMode> = ContextFn<
  BuildMode,
  GraphQLType | ContextModel<BuildMode, any>
>
export const toContextFn = <BuildMode>(
  type:
    | ModelType<BuildMode>
    | ModelBuilder<BuildMode, any>
    | ContextFn<BuildMode, GraphQLType>,
): ToContextFnResult<BuildMode> => {
  if (typeof type === 'function') return type
  if (isType(type)) return () => type
  return buildMode => buildMode.getModel(type.name)
}

export type TypeCondition = 'nonnull' | 'list' | 'none'

export const toList = <Type extends GraphQLType = GraphQLType>(type: Type) =>
  GraphQLNonNull(GraphQLList(GraphQLNonNull(type))) as Type

export const conditionalList = <Type extends GraphQLType>(
  type: Type,
  isList: boolean,
) => (isList ? toList(type) : type)

export const conditionalNonNull = <Type extends GraphQLType>(
  type: Type,
  nonNull: boolean,
) => (nonNull ? GraphQLNonNull(type) : type)

export const conditionalType = <Type extends GraphQLType>(
  type: Type,
  condition: TypeCondition,
) =>
  condition === 'list'
    ? toList(type)
    : condition === 'nonnull'
    ? GraphQLNonNull(type)
    : type

export const memoizeContextModel = <BuildMode = any, Result = any, Type = any>(
  fn: (buildModeModel: ContextModel<BuildMode, Type, any>) => Result,
) =>
  memoize(
    fn,
    (buildModeModel: ContextModel<BuildMode, Type>) => buildModeModel.id,
  )

export const reduceContextFields = <
  BuildMode,
  Type extends Record<string, any>
>(
  buildModeModel: ContextModel<BuildMode, Type>,
  base: Type = null,
  reducer: (
    memo: Type,
    attr: AttributeBuilder<BuildMode, Type, any>,
    type: GraphQLType,
    field: ModelType<BuildMode>,
  ) => Type,
) =>
  buildModeModel
    .getFields()
    .reduce(
      (memo: Type, attr) =>
        reducer(
          memo,
          attr,
          buildType(attr, buildModeModel.buildMode),
          attr.field(buildModeModel.buildMode),
        ),
      base || {},
    )

export const createPageType = <Type>(page: PageData, nodes: Type[]) => ({
  page,
  nodes,
})

interface ContextModelFieldFnConfig {
  iterator: string
  condition: TypeCondition
}
export const createModelFieldFn = <BuildMode>(
  configFn: (
    buildModeModel: ContextModel<BuildMode, any>,
  ) => ContextModelFieldFnConfig,
) => (buildModeModel: ContextModel<BuildMode, any>) => {
  const { iterator, condition = 'none' } = configFn(buildModeModel)
  return {
    subscribe: () => buildModeModel.buildMode.pubSub.asyncIterator(iterator),
    resolve: ({ node }) => node,
    type: conditionalType(
      buildModeModel.getType() as GraphQLInputType,
      condition,
    ),
  }
}

export const createInputType = <
  BuildMode,
  Key extends keyof Names,
  Name extends keyof Names[Key]
>(
  field: DataType,
  nameType: Key,
  name: Name,
) =>
  memoizeContextModel(
    buildModeModel =>
      new GraphQLInputObjectType({
        name: (buildModeModel.names[nameType] as Record<any, string>)[
          name
        ] as string,
        fields: () =>
          buildModeModel.dataFields(field) as GraphQLInputFieldConfigMap,
      }),
  )
