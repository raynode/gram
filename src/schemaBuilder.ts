import { DateTime } from '@saeris/graphql-scalars'
import {
  GraphQLFieldConfigMap,
  GraphQLFieldResolver,
  GraphQLID,
  GraphQLInputFieldConfigMap,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLSchema,
  GraphQLSchemaConfig,
  GraphQLString,
  GraphQLType,
  isType,
} from 'graphql'
import { PubSub } from 'graphql-subscriptions'
import { GraphQLJSONObject } from 'graphql-type-json'
import { filter, forEach, reduce } from 'lodash'
import { v4 as uuid } from 'uuid'

import { createModelBuilder } from './modelBuilder'
import { createFilterStrategy, defaultMiddlewares } from './strategies/filter'
import {
  ContextModel,
  ContextMutator,
  FieldDefinition,
  FilterMiddleware,
  GenericGraphQLType,
  ListType,
  ModelBuilder,
  NodeType,
  PageData,
  QueryTypeDefinition,
  SchemaBuilder,
  Service,
  WithContext,
  Wrapped,
} from './types'
import { SCHEMABUILDER } from './types/constants'
import { isFieldDefinition } from './types/guards'

import { addModel, createBuild } from './createBuild'

const wrapContext = <BuildMode>(
  buildMode: BuildMode | null,
  scalars: Record<string, GraphQLScalarType>,
  models: Models<BuildMode>,
  filters: FilterMiddleware[],
  pubSub: PubSub,
): Wrapped<BuildMode> => {
  const buildModeModels: Record<string, ContextModel<BuildMode, any>> = {}
  return {
    id: uuid(),
    buildMode,
    filterStrategy: createFilterStrategy(filters),
    addModel: (name, model) => (buildModeModels[name] = model),
    getBaseModel: name => models[name],
    getModel: name => buildModeModels[name],
    getScalar: name => scalars[name],
    pubSub,
  }
}

const addNodeAttrs = <BuildMode>(model: ModelBuilder<BuildMode, any>) => {
  model.attr('id', GraphQLID)
  model.attr('createdAt', buildMode => buildMode.getScalar('DateTime'))
  model.attr('updatedAt', buildMode => buildMode.getScalar('DateTime'))
  model.attr('deletedAt', buildMode => buildMode.getScalar('DateTime'))
  return model
}

const createBaseModels = <BuildMode>() => {
  const node = createModelBuilder<BuildMode, NodeType>(
    'Node',
    {},
  ).setInterface()
  const page = createModelBuilder<BuildMode, PageData>('Page', {})
  const list = createModelBuilder<BuildMode, ListType<NodeType>>(
    'List',
    {},
  ).setInterface()

  addNodeAttrs(node)

  page.attr('page', GraphQLInt)
  page.attr('limit', GraphQLInt)
  page.attr('offset', GraphQLInt)

  list.attr('page', page)
  list.attr('nodes', node).isList()

  return {
    Node: node,
    Page: page,
    List: list,
  }
}

type Models<BuildMode> = Record<string, ModelBuilder<BuildMode, any, any>>

const setup = <BuildMode>(
  models: Models<BuildMode>,
  scalars: Record<string, GraphQLScalarType>,
  buildMode: BuildMode | null,
  filters: FilterMiddleware[],
  pubSub: PubSub = new PubSub(),
) => {
  const wrapped = wrapContext(buildMode, scalars, models, filters, pubSub)
  forEach(models, model => model.setup(wrapped))
  return wrapped
}

export const createSchema = (definition: FieldDefinition) => {
  const schema: GraphQLSchemaConfig = {
    query: new GraphQLObjectType({
      name: 'Query',
      fields: definition.query,
    }),
  }
  if (Object.keys(definition.mutation).length)
    schema.mutation = new GraphQLObjectType({
      name: 'Mutation',
      fields: definition.mutation,
    })
  if (Object.keys(definition.subscription).length)
    schema.subscription = new GraphQLObjectType({
      name: 'Subscription',
      fields: definition.subscription,
    })
  return new GraphQLSchema(schema)
}

export const createSchemaBuilder = <BuildMode = any, QueryContext = any>() => {
  const models: Models<BuildMode> = createBaseModels<BuildMode>()
  const scalars: Record<string, GraphQLScalarType> = { DateTime }
  const filters = defaultMiddlewares
  let externalPubSub: PubSub = null
  const queryDefinitions: Array<
    WithContext<BuildMode, QueryTypeDefinition<BuildMode, any, QueryContext>>
  > = []

  const builder: SchemaBuilder<BuildMode, QueryContext> = {
    type: SCHEMABUILDER,
    models,
    model: <Type, GQLType = Type>(
      modelName: string,
      service: Service<Type, GQLType, QueryContext>,
    ) => {
      const model = createModelBuilder<BuildMode, Type, GQLType>(
        modelName,
        service || {},
      )
      models[modelName] = model
      return model.interface('Node')
    },
    interface: <Type>(
      interfaceName: string,
      service: Service<Type, Type, QueryContext>,
    ) => {
      const model = createModelBuilder<BuildMode, Type>(
        interfaceName,
        service || {},
      )
      models[interfaceName] = model
      model.setInterface()
      return model
    },
    build: (buildMode: BuildMode | FieldDefinition = null) =>
      isFieldDefinition(buildMode)
        ? createSchema(buildMode)
        : builder.createBuild(buildMode).toSchema(),
    createBuild: buildMode => {
      const pubSub = externalPubSub || new PubSub()
      const wrapped = setup(models, scalars, buildMode, filters, pubSub)

      // build all interfaces
      filter(models, model => model.isInterface()).forEach(model =>
        model.build(wrapped),
      )

      const build = addModel(
        createBuild<BuildMode, QueryContext>(buildMode),
        wrapped,
      )
      forEach(models, build.addModel)
      forEach(scalars, scalar => build.addType(scalar.toString(), 'scalar'))

      forEach(queryDefinitions, queryDefinition => {
        const { name, args, type, resolver } =
          typeof queryDefinition === 'function'
            ? queryDefinition(wrapped)
            : queryDefinition
        const typeName = isType(type)
          ? type.toString()
          : typeof type === 'string'
          ? type
          : type.name
        build.addQuery(name, { args, type: typeName }, resolver)
      })
      return build
    },
    setScalar: (key, type) => {
      scalars[key] = type
      return type
    },
    getScalar: key => scalars[key],
    setPubSub: (pubSub: PubSub) => {
      externalPubSub = pubSub
      return builder
    },
    addFilter: (check, filter) => {
      filters.push([check, filter])
      return builder
    },
    addQuery: definition => {
      queryDefinitions.push(definition)
      return builder
    },
  }

  return builder
}
