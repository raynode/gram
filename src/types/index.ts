import {
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLFieldResolver,
  GraphQLInputFieldConfigMap,
  GraphQLInputObjectType,
  GraphQLInterfaceType,
  GraphQLList,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLSchema,
  GraphQLType,
  Thunk,
} from 'graphql'
import { PubSub } from 'graphql-subscriptions'

import { IFieldResolver } from 'graphql-tools'

import { WithAddModel } from '../createBuild/extension-model'

import { Names } from '../strategies/naming'
import {
  ATTRIBUTEBUILDER,
  MODELBUILDER,
  SCHEMABUILDER,
} from '../types/constants'
import { Service } from '../types/service'
export * from '../types/service'

export interface Builder {
  type: string
}

export type Attributes<BuildMode, Type> = Record<
  string,
  AttributeBuilder<BuildMode, any, Type>
>

export type Fields = Thunk<GraphQLFieldConfigMap<any, any>>
export type FilterFn = (
  name: string,
  type: GraphQLType,
  list: GraphQLList<any>,
) => Record<string, { type: GraphQLType }>
export type FilterCheckFn = (
  type: GraphQLType,
  required: boolean,
  list: boolean,
  baseType: GraphQLType,
) => boolean
export type FilterMiddleware = [FilterCheckFn, FilterFn]

export type FilterStrategy = <Type extends GraphQLType = GraphQLType>(
  inputType: Type | ContextModel<any, any>,
  inputName?: string,
) => Record<string, { type: GraphQLType }>

export type DataType =
  | 'create'
  | 'filter'
  | 'data'
  | 'list'
  | 'page'
  | 'where'
  | 'order'

export type FieldTypes = 'GraphQL' | 'Model' | 'All'
export type GenericGraphQLType = 'Date'
export type FieldDefinition = Record<
  'query' | 'mutation' | 'subscription',
  GraphQLFieldConfigMap<any, any>
>

export interface ContextModel<BuildMode, Type, GQLType = Type> {
  addField: <AttributeType>(
    field: AttributeBuilder<BuildMode, Type, AttributeType>,
  ) => void
  baseFilters: () => GraphQLInputFieldConfigMap
  buildMode: Wrapped<BuildMode>
  dataFields: (
    type: DataType,
  ) => GraphQLInputFieldConfigMap | GraphQLFieldConfigMap<any, any>
  getFields: () => Array<AttributeBuilder<BuildMode, Type, any>>
  getListType: () => GraphQLType
  getType: () => GraphQLType
  id: string
  isInterface: () => boolean
  name: string
  names: Names
  service: Service<Type, GQLType>
  setInterface: () => void
  visibility: ModelVisibility
}

export interface Wrapped<BuildMode> {
  id: string
  getBaseModel: <Type, GQLType>(
    name: string,
  ) => ModelBuilder<BuildMode, Type, GQLType>
  getModel: <Type, GQLType>(
    name: string,
  ) => ContextModel<BuildMode, Type, GQLType>
  addModel: <Type, GQLType>(
    name: string,
    model: ContextModel<BuildMode, Type, GQLType>,
  ) => void
  filterStrategy: FilterStrategy
  buildMode: BuildMode | null
  getScalar: (key: string) => GraphQLScalarType
  pubSub: PubSub
}

export type ContextModelFn<Result> = <BuildMode>(
  buildModeModel: ContextModel<BuildMode, any, any>,
) => Result

export type ContextFn<BuildMode, Result = boolean> = (
  buildMode: Wrapped<BuildMode>,
) => Result
export type ContextMutator<BuildMode, Type, GQLType> = (
  model: ContextModel<BuildMode, Type, GQLType>,
  buildMode: Wrapped<BuildMode>,
) => void
export type ContextModelFieldFn<Type, GQLType> = <BuildMode>(
  buildModeModel: ContextModel<BuildMode, Type, GQLType>,
) => Type

export type WithContext<BuildMode, Type> = Type | ContextFn<BuildMode, Type>

export type ModelType<BuildMode> = GraphQLType | ContextModel<BuildMode, any>

export interface QueryTypeDefinition<
  BuildMode,
  Type,
  QueryContext,
  Args extends Record<string, any> = any
> {
  name: string
  args?: Args
  type: string | GraphQLType | ModelBuilder<BuildMode, Type>
  resolver: GraphQLFieldResolver<null, QueryContext, Args>
}
// GraphQLFieldResolver<TSource, TContext, TArgs>

export interface SchemaBuilder<BuildMode, QueryContext = any> extends Builder {
  build: (buildMode?: BuildMode | FieldDefinition) => GraphQLSchema
  interface: <Type>(
    interfaceName: string,
    service?: Service<Type>,
  ) => ModelBuilder<BuildMode, Type>
  model: <Type, GQLType = Type>(
    modelName: string,
    service?: Service<Type, GQLType>,
  ) => ModelBuilder<BuildMode, Type, GQLType>
  fields: (buildMode?: BuildMode) => FieldDefinition
  models: Record<string, ModelBuilder<BuildMode, any>>
  type: typeof SCHEMABUILDER
  setScalar: <Type extends GraphQLScalarType>(key: string, type: Type) => Type
  getScalar: (key: string) => GraphQLScalarType
  addFilter: (check: FilterCheckFn, filter: FilterFn) => this
  createBuild: (buildMode?: BuildMode) => WithAddModel<BuildMode, QueryContext>
  addQuery: <Type>(
    definition: WithContext<
      BuildMode,
      QueryTypeDefinition<BuildMode, Type, QueryContext>
    >,
  ) => this
  setPubSub: (pubSub: PubSub) => this
  // addMutation: () => this
}

export type GraphQLResolverMap<GQLType, Attrs extends string = string> = Record<
  Attrs,
  IFieldResolver<GQLType, any>
>

export interface ModelBuilder<BuildMode, Type, GQLType = Type> extends Builder {
  attr: <AttributeType>(
    attributeName: string,
    type:
      | ModelType<BuildMode>
      | ModelBuilder<BuildMode, any>
      | ContextFn<BuildMode, GraphQLType>,
  ) => AttributeBuilder<BuildMode, Type, AttributeType>
  resolve: <Attrs extends string>(
    resolver: ContextFn<BuildMode, Record<string, IFieldResolver<GQLType, any>>>,
  ) => this
  getResolver: (
    buildMode: Wrapped<BuildMode>,
  ) => Record<string, IFieldResolver<GQLType, any>>
  build: (
    buildMode: Wrapped<BuildMode>,
  ) => ContextModel<BuildMode, Type, GQLType>
  buildMode: (
    buildModeMutation: ContextMutator<BuildMode, Type, GQLType>,
  ) => this
  getAttributes: () => Attributes<BuildMode, Type>
  getInterfaces: () => string[]
  getListType: () => GraphQLType | ModelBuilder<BuildMode, any>
  getVisibility: () => ModelVisibility
  getService: () => Service<Type, GQLType>
  interface: (model: string) => this
  isInterface: () => boolean
  listType: (model: ModelBuilder<BuildMode, any>) => this
  name: string
  setInterface: () => this
  setup: ContextFn<BuildMode, void>
  type: typeof MODELBUILDER
}

export interface AttributeBuilder<BuildMode, Type, AttributeType>
  extends Builder {
  name: string
  field: ContextFn<BuildMode, GraphQLType | ContextModel<BuildMode, Type, any>>
  nullable: boolean
  listType: boolean
  resolve: (resolver: GraphQLFieldResolver<Type, any>) => this
  isList: (isList?: boolean) => this
  isNotNullable: (isNotNullable?: boolean) => this
  // buildModeType: ContextFn<BuildMode, ModelType<BuildMode>>
  build: ContextFn<BuildMode, GraphQLFieldConfig<any, any>>
  type: typeof ATTRIBUTEBUILDER
}

export interface ModelVisibility {
  createMutation: boolean
  deleteMutation: boolean
  findManyQuery: boolean
  findOneQuery: boolean
  updateMutation: boolean
  createSubscription: boolean
  updateSubscription: boolean
  deleteSubscription: boolean
}
