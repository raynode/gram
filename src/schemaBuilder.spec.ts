import {
  graphql,
  GraphQLFloat,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLString,
  printSchema,
} from 'graphql'

import { toList } from 'utils'
import { createSchemaBuilder } from './schemaBuilder'
import { ModelBuilder, NodeType, SchemaBuilder } from './types'

import {
  Account,
  Accounts,
  db,
  findMany,
  Nodes,
  Page,
  Paged,
  reset,
  User,
  Users,
} from '__mocks__/database'

describe('schemaBuilder', () => {
  let builder: SchemaBuilder<number>

  beforeEach(() => {
    builder = createSchemaBuilder<number>()
  })

  it('should have basic methods', () => {
    expect(builder).toHaveProperty('model')
    expect(typeof builder.model).toBe('function')
  })

  it('should render an empty Schema', () => {
    expect(printSchema(builder.build(0))).toMatchSnapshot()
  })
})

describe('example', () => {
  let builder: SchemaBuilder<number>

  beforeAll(() => {
    reset()
    builder = createSchemaBuilder<number>()
  })

  it('should add User to the Schema', () => {
    const user = builder.model('User', Users)
    user.attr('name', GraphQLString)
    user.attr('friends', user).isList()
    const schema = builder.build(0)
    expect(printSchema(schema)).toMatchSnapshot()
  })

  it('should add account to the model', () => {
    const account = builder.model('Account', Accounts)

    // will add input types for "STRING"
    account.attr('name', GraphQLString)
    account.attr('amount', GraphQLFloat)
    // will add input types for "REFERENCE"
    account
      .attr('user', builder.models.User)
      .resolve(account =>
        Users.findOne({ where: { id: account.userId }, order: null }),
      )
      .isNotNullable()
    // add accounts to the user model
    builder.models.User.attr('accounts', account)
      .resolve(user => findMany({ userId: user.id }, null))
      .isList()

    expect(printSchema(builder.build(0))).toMatchSnapshot()
  })

  it('should be able read from the schema', async () => {
    const schema = builder.build(0)

    const query = `{
      getUser(where: {
        id: "1",
      }) {
        name
      }
    }`

    const { data } = await graphql(schema, query, null)
    expect(data).toMatchSnapshot()
  })

  it('should be able to create an instance', async () => {
    const schema = builder.build(0)

    const query = `mutation {
      createUser(data: {
        name: "New Test-User",
      }) {
        name
      }
    }`

    const { data, errors } = await graphql(schema, query, null)
    expect(data).toHaveProperty('createUser')
    expect(data.createUser).toHaveProperty('name')
    expect(data.createUser.name).toEqual('New Test-User')
  })

  it('should find multiple things', async () => {
    const schema = builder.build(0)

    // where is required by may be an empty thing
    const query = `{
      getUsers(where: {}) {
        nodes {
          name
        }
      }
    }`

    const { data } = await graphql(schema, query, null)
    expect(data).toMatchSnapshot()
  })

  it('should be able to delete instances', async () => {
    const schema = builder.build(0)

    const query = `mutation {
      deleteUsers(where: { id: "1" }) {
        name
      }
    }`

    const { data } = await graphql(schema, query, null)
    expect(db).not.toHaveProperty('1')
  })

  it('should create an account for the user', async () => {
    const schema = builder.build(0)

    const {
      data: {
        getUser: { id },
      },
    } = await graphql(
      schema,
      `
        {
          getUser(where: { name: "New Test-User" }) {
            id
          }
        }
      `,
      null,
    )

    const query = `mutation($id: ID!) {
      createAccount(data: {
        name: "New Account"
        user: { id: $id }
      }) {
        id
        name
        amount
        user {
          name
        }
      }
    }`

    const { data } = await graphql(schema, query, null, null, { id })
    expect(data).toHaveProperty('createAccount')
    expect(data.createAccount).toHaveProperty('id')
    const { id: accountId, user } = data.createAccount
    expect(db).toHaveProperty(accountId)
    expect(user).toHaveProperty('name')
    expect(user.name).toEqual((db[id] as User).name)
  })

  it('should find the new account in the user as well', async () => {
    const schema = builder.build(0)

    const { data, errors } = await graphql(
      schema,
      `
        {
          getUser(where: { name: "New Test-User" }) {
            name
            accounts {
              name
            }
          }
        }
      `,
    )

    expect(data).toHaveProperty('getUser')
    expect(data.getUser).toHaveProperty('name')
    expect(data.getUser).toHaveProperty('accounts')
    expect(data.getUser.name).toEqual('New Test-User')
    expect(data.getUser.accounts).toHaveLength(1)
    expect(data.getUser.accounts[0].name).toEqual('New Account')
  })
})
