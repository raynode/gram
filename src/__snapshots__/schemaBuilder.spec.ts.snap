// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example should add User to the Schema 1`] = `
"input CreateUserData {
  name: String
  friends: UserWhere
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

type Mutation {
  createUser(data: CreateUserData!): User
  updateUser(data: UpdateUserData!, where: UserWhere!): [User!]!
  deleteUsers(where: UserWhere!): [User!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getUser(where: UserWhere!, order: UserSortOrder): User
  getUsers(order: UserSortOrder, page: UserPage, where: UserWhere!): Users
}

type Subscription {
  onCreateUser: User!
  onUpdateUser: [User!]!
  onDeleteUsers: [User!]!
}

input UpdateUserData {
  name: String
  friends: UserWhere
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type User implements Node {
  name: String
  friends: [User!]!
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input UserFilter {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input UserPage {
  limit: Int
  offset: Int
}

type Users implements List {
  page: Page
  nodes: [User!]!
}

enum UserSortOrder {
  name_ASC
  name_DESC
  friends_ASC
  friends_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input UserWhere {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}
"
`;

exports[`example should add account to the model 1`] = `
"type Account implements Node {
  name: String
  amount: Float
  user: User!
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input AccountFilter {
  AND: [AccountFilter!]
  OR: [AccountFilter!]
  NOT: AccountFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  amount: Float
  amount_not: Float
  amount_lt: Float
  amount_gt: Float
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input AccountPage {
  limit: Int
  offset: Int
}

type Accounts implements List {
  page: Page
  nodes: [Account!]!
}

enum AccountSortOrder {
  name_ASC
  name_DESC
  amount_ASC
  amount_DESC
  user_ASC
  user_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input AccountWhere {
  AND: [AccountFilter!]
  OR: [AccountFilter!]
  NOT: AccountFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  amount: Float
  amount_not: Float
  amount_lt: Float
  amount_gt: Float
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input CreateAccountData {
  name: String
  amount: Float
  user: UserWhere!
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input CreateUserData {
  name: String
  friends: UserWhere
  accounts: AccountWhere
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

type Mutation {
  createUser(data: CreateUserData!): User
  updateUser(data: UpdateUserData!, where: UserWhere!): [User!]!
  deleteUsers(where: UserWhere!): [User!]!
  createAccount(data: CreateAccountData!): Account
  updateAccount(data: UpdateAccountData!, where: AccountWhere!): [Account!]!
  deleteAccounts(where: AccountWhere!): [Account!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getUser(where: UserWhere!, order: UserSortOrder): User
  getUsers(order: UserSortOrder, page: UserPage, where: UserWhere!): Users
  getAccount(where: AccountWhere!, order: AccountSortOrder): Account
  getAccounts(order: AccountSortOrder, page: AccountPage, where: AccountWhere!): Accounts
}

type Subscription {
  onCreateUser: User!
  onUpdateUser: [User!]!
  onDeleteUsers: [User!]!
  onCreateAccount: Account!
  onUpdateAccount: [Account!]!
  onDeleteAccounts: [Account!]!
}

input UpdateAccountData {
  name: String
  amount: Float
  user: UserWhere
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input UpdateUserData {
  name: String
  friends: UserWhere
  accounts: AccountWhere
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type User implements Node {
  name: String
  friends: [User!]!
  accounts: [Account!]!
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input UserFilter {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input UserPage {
  limit: Int
  offset: Int
}

type Users implements List {
  page: Page
  nodes: [User!]!
}

enum UserSortOrder {
  name_ASC
  name_DESC
  friends_ASC
  friends_DESC
  accounts_ASC
  accounts_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input UserWhere {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}
"
`;

exports[`example should be able read from the schema 1`] = `
Object {
  "getUser": Object {
    "name": "test",
  },
}
`;

exports[`example should find multiple things 1`] = `
Object {
  "getUsers": Object {
    "nodes": Array [
      Object {
        "name": "test",
      },
      Object {
        "name": "New Test-User",
      },
    ],
  },
}
`;

exports[`schemaBuilder should render an empty Schema 1`] = `
"scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}
"
`;
